### PR TITLE
Add subnetCount to pod_ip_metrics.go

### PR DESCRIPTION
**Pull request by yiningou**
Sorry for the confusion, The pr: https://github.com/GoogleCloudPlatform/netd/pull/205 was somehow closed when I tried to create a new pr #211 to change the conntrack.go, tried to reopen it but failed. Open a new pr with the two commits on pr #205 .

/assign jingyuanliang
/assign MrHohn

### CODE REVIEW

**Review by MrHohn — APPROVED** [review]
/lgtm
/approve
/hold

File: pkg/metrics/collector/pod_ip_metrics.go [hunk 1]
```diff
@@ -139,7 +158,26 @@ func init() {
...
+	if err != nil {
+		return nil, fmt.Errorf("error creating in-cluster config: %v", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("error creating clientset: %v", err)
+	}
+
+	nodeName := os.Getenv("HOSTNAME")
```
Comment (jingyuanliang) [inline pkg/metrics/collector/pod_ip_metrics.go:171]: `os.Hostname()` or `k8s.io/component-helpers/node/util.GetHostname("")` (this runs some canonicalization)

File: pkg/metrics/collector/pod_ip_metrics.go [hunk 2]
```diff
@@ -266,6 +304,68 @@ func (c *podIPMetricsCollector) fillBuckets(diff uint64) {
...
+	// this checks that we are not overflowing an int64
+	if bits-ones >= 64 {
+		return math.MaxUint64, nil
+	}
+	max := uint64(1) << uint(bits-ones)
+	max--
+	if subnet.IP.To4() != nil {
+		// Don't use the IPv4 network's broadcast address
+		if max == 0 {
+			return 0, fmt.Errorf("subnet includes only the network and broadcast addresses")
```
Comment (jingyuanliang) [inline pkg/metrics/collector/pod_ip_metrics.go:327]: This error message doesn't look right. The code works as:

IPv4, /32: return 0, err "subnet includes only the network and broadcast addresses"
IPv4, /31: return 0, nil
IPv4, /30: return 2, nil
IPv6, /128: return 0, nil
IPv6, /127: return 1, nil
IPv6, /126: return 3, nil

File: pkg/metrics/collector/pod_ip_metrics.go [hunk 3]
```diff
@@ -266,6 +304,68 @@ func (c *podIPMetricsCollector) fillBuckets(diff uint64) {
...
+	}
+}
+
+func (c *podIPMetricsCollector) calculateAssignedIP() error {
+	node, err := c.clientset.CoreV1().Nodes().Get(context.Background(), c.nodeName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting node %s: %v", c.nodeName, err)
+	}
+
+	var allCIDRs []string
```
Comment (jingyuanliang) [inline pkg/metrics/collector/pod_ip_metrics.go:348]: I would write something like:

```
podCIDRs := node.Spec.PodCIDRs
if len(podCIDRs) == 0 && node.Spec.PodCIDR != "" {
    podCIDRs = []string{node.Spec.PodCIDR}
}
```

File: pkg/metrics/collector/pod_ip_metrics.go [hunk 4]
```diff
@@ -266,6 +304,68 @@ func (c *podIPMetricsCollector) fillBuckets(diff uint64) {
...
+		}
+	}
+	for _, podCIDR := range allCIDRs {
+		_, subnet, err := net.ParseCIDR(podCIDR)
+		if err != nil {
+			return fmt.Errorf("error parsing podCIDR %s: %v", podCIDR, err)
+		}
+		totalIP, err := c.countIPsFromRange(subnet)
+		if err != nil {
+			return fmt.Errorf("error calculating total IPs for subnet %s: %v", subnet.IP.String(), err)
```
Comment (jingyuanliang) [inline pkg/metrics/collector/pod_ip_metrics.go:362]: %s calls String() automatically.

File: pkg/metrics/collector/pod_ip_metrics.go [hunk 5]
```diff
@@ -139,7 +158,29 @@ func init() {
...
+	if err != nil {
+		return nil, fmt.Errorf("error creating clientset: %v", err)
+	}
+
+	nodeName, err := os.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("error getting hostname: %v", err)
+	}
+	if nodeName == "" {
+		return nil, fmt.Errorf("HOSTNAME environment variable not set")
```
Comment (jingyuanliang) [inline pkg/metrics/collector/pod_ip_metrics.go:176]: https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/os/sys_linux.go it's not based on env. Will it return empty anyway (I'm not sure)?

File: pkg/metrics/collector/pod_ip_metrics.go (hunk 4 above)
Comment (yiningou) [reply pkg/metrics/collector/pod_ip_metrics.go]: Done.

File: pkg/metrics/collector/pod_ip_metrics.go (hunk 3 above)
Comment (yiningou) [reply pkg/metrics/collector/pod_ip_metrics.go]: Done.

File: pkg/metrics/collector/pod_ip_metrics.go (hunk 2 above)
Comment (yiningou) [reply pkg/metrics/collector/pod_ip_metrics.go:330]: Done.

File: pkg/metrics/collector/pod_ip_metrics.go (hunk 1 above)
Comment (yiningou) [reply pkg/metrics/collector/pod_ip_metrics.go]: Done.

File: pkg/metrics/collector/pod_ip_metrics_test.go [hunk 6]
```diff
@@ -269,3 +285,50 @@ func TestSetupDirectoryWatcher(t *testing.T) {
...
+			podCIDRs:              []string{"10.0.0.0/24", "2600:1900::/125"},
+			wantAssignedIPv4Count: 254,
+			wantAssignedIPv6Count: 7,
+		},
+		{
+			desc:                  "Two same IPv4 addresses and no IPv6 address",
+			podCIDR:               "10.0.0.0/24",
+			podCIDRs:              []string{"10.0.0.0/24"},
+			wantAssignedIPv4Count: 254,
+			wantAssignedIPv6Count: 0,
```
Comment (jingyuanliang) [inline pkg/metrics/collector/pod_ip_metrics_test.go:309]: Can we add a few test cases on missing podCIDR / podCIDRs or invalid ones?

File: pkg/metrics/collector/pod_ip_metrics.go [hunk 7]
```diff
@@ -266,6 +304,77 @@ func (c *podIPMetricsCollector) fillBuckets(diff uint64) {
...
+		return fmt.Errorf("error getting node %s: %v", c.nodeName, err)
+	}
+
+	podCIDRs := node.Spec.PodCIDRs
+	if len(podCIDRs) == 0 && node.Spec.PodCIDR != "" {
+		podCIDRs = []string{node.Spec.PodCIDR}
+	}
+	if len(podCIDRs) == 0 {
+		return fmt.Errorf("both podCIDR and podCIDRs are empty")
+	}
```
Comment (jingyuanliang) [inline pkg/metrics/collector/pod_ip_metrics.go:356]: Ah you're checking this too. Maybe this for one less evaluation?

```
	podCIDRs := node.Spec.PodCIDRs
	if len(podCIDRs) == 0 {
		if node.Spec.PodCIDR == "" {
			return fmt.Errorf("both podCIDR and podCIDRs are empty")
		} else {
			podCIDRs = []string{node.Spec.PodCIDR}
		}
	}
```

Comment (yiningou) [reply pkg/metrics/collector/pod_ip_metrics.go:356]: Done.

**Pull request merged**